{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "gpuClass": "standard"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/lauraluebbert/kallistoXpalmdb/blob/main/kallistoXpalmDB_tutorial.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# kallisto X PalmDB tutorial\n",
        "\n",
        "This is a simple pre-release tutorial introducig the new `--aa` option of [kallisto](https://www.nature.com/articles/s41587-021-00870-2) to generate count matrices from scRNAseq data using an amino acid referece. Here, we will use this option in combination with the [PalmDB database](https://www.nature.com/articles/s41586-021-04332-2) to find viral sequences in [Rhesus macaques infected with Ebolavirus](https://pubmed.ncbi.nlm.nih.gov/33159858/)."
      ],
      "metadata": {
        "id": "zKAMqEdz4R3n"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install ffq (https://github.com/pachterlab/ffq) and gget (https://github.com/pachterlab/gget) for simplified data download\n",
        "!pip install -q anndata gget ffq\n",
        "\n",
        "import anndata\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "def nd(arr):\n",
        "    \"\"\"\n",
        "    Function to transform numpy matrix to nd array.\n",
        "    \"\"\"\n",
        "    return np.asarray(arr).reshape(-1)"
      ],
      "metadata": {
        "id": "VkA-4EjdI5ym",
        "outputId": "970cf934-a44f-4cf3-e3d2-4fdf08a06778",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m103.0/103.0 kB\u001b[0m \u001b[31m2.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.2/2.2 MB\u001b[0m \u001b[31m12.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m25.2/25.2 MB\u001b[0m \u001b[31m33.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m17.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Install kallisto from the dev branch"
      ],
      "metadata": {
        "id": "5ddSnwmkAw6k"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next, we will install the kallisto bustools wrapper [`kb-python`](https://github.com/pachterlab/kb_python) and compile kallisto from the development branch (after the official release, this will become just `pip install kb-python`):"
      ],
      "metadata": {
        "id": "Y9uFetKXIzSO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install kb_python, kallisto and bustools from dev branch\n",
        "# After the official release, everything in this cell will become 'pip install kb-python'\n",
        "!pip install -q git+https://github.com/pachterlab/kb_python.git@dlist\n",
        "\n",
        "!git clone https://github.com/pachterlab/kallisto-D.git\n",
        "!mkdir kallisto-D/build && cd kallisto-D/build && cmake .. && make && make install\n",
        "\n",
        "!git clone https://github.com/BUStools/bustools.git && cd bustools && git checkout dlist\n",
        "!mkdir bustools/build && cd bustools/build && cmake .. && make && make install"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_icuVIGtIMDQ",
        "outputId": "067483e8-f5ff-4da5-ee87-0d85d3a2fa71"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.8/4.8 MB\u001b[0m \u001b[31m38.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m45.2/45.2 MB\u001b[0m \u001b[31m6.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.0/2.0 MB\u001b[0m \u001b[31m70.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m53.3/53.3 kB\u001b[0m \u001b[31m6.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m20.0/20.0 MB\u001b[0m \u001b[31m64.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m88.2/88.2 kB\u001b[0m \u001b[31m10.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.1/1.1 MB\u001b[0m \u001b[31m63.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m63.5/63.5 kB\u001b[0m \u001b[31m7.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Building wheel for kb-python (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for loompy (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for ngs-tools (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for umap-learn (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for numpy-groupies (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for session-info (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for pynndescent (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Cloning into 'kallisto-D'...\n",
            "remote: Enumerating objects: 7618, done.\u001b[K\n",
            "remote: Counting objects: 100% (419/419), done.\u001b[K\n",
            "remote: Compressing objects: 100% (208/208), done.\u001b[K\n",
            "remote: Total 7618 (delta 297), reused 315 (delta 211), pack-reused 7199\u001b[K\n",
            "Receiving objects: 100% (7618/7618), 8.56 MiB | 9.98 MiB/s, done.\n",
            "Resolving deltas: 100% (5049/5049), done.\n",
            "-- The C compiler identification is GNU 9.4.0\n",
            "-- The CXX compiler identification is GNU 9.4.0\n",
            "-- Detecting C compiler ABI info\n",
            "-- Detecting C compiler ABI info - done\n",
            "-- Check for working C compiler: /usr/bin/cc - skipped\n",
            "-- Detecting C compile features\n",
            "-- Detecting C compile features - done\n",
            "-- Detecting CXX compiler ABI info\n",
            "-- Detecting CXX compiler ABI info - done\n",
            "-- Check for working CXX compiler: /usr/bin/c++ - skipped\n",
            "-- Detecting CXX compile features\n",
            "-- Detecting CXX compile features - done\n",
            "-- Performing Test COMPILER_SUPPORTS_CXX17\n",
            "-- Performing Test COMPILER_SUPPORTS_CXX17 - Success\n",
            "\u001b[0mshared build\u001b[0m\n",
            "-- Performing Test CMAKE_HAVE_LIBC_PTHREAD\n",
            "-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed\n",
            "-- Looking for pthread_create in pthreads\n",
            "-- Looking for pthread_create in pthreads - not found\n",
            "-- Looking for pthread_create in pthread\n",
            "-- Looking for pthread_create in pthread - found\n",
            "-- Found Threads: TRUE  \n",
            "-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version \"1.2.11\") \n",
            "-- Configuring done\n",
            "-- Generating done\n",
            "-- Build files have been written to: /content/kallisto-D/build\n",
            "[  3%] \u001b[34m\u001b[1mCreating directories for 'bifrost'\u001b[0m\n",
            "[  7%] \u001b[34m\u001b[1mNo download step for 'bifrost'\u001b[0m\n",
            "[ 11%] \u001b[34m\u001b[1mNo update step for 'bifrost'\u001b[0m\n",
            "[ 15%] \u001b[34m\u001b[1mNo patch step for 'bifrost'\u001b[0m\n",
            "[ 19%] \u001b[34m\u001b[1mPerforming configure step for 'bifrost'\u001b[0m\n",
            "-- The C compiler identification is GNU 9.4.0\n",
            "-- The CXX compiler identification is GNU 9.4.0\n",
            "-- Detecting C compiler ABI info\n",
            "-- Detecting C compiler ABI info - done\n",
            "-- Check for working C compiler: /usr/bin/cc - skipped\n",
            "-- Detecting C compile features\n",
            "-- Detecting C compile features - done\n",
            "-- Detecting CXX compiler ABI info\n",
            "-- Detecting CXX compiler ABI info - done\n",
            "-- Check for working CXX compiler: /usr/bin/c++ - skipped\n",
            "-- Detecting CXX compile features\n",
            "-- Detecting CXX compile features - done\n",
            "\u001b[0mCompilation architecture: native\u001b[0m\n",
            "\u001b[0mBuild type: Release\u001b[0m\n",
            "\u001b[0mMaximum k-mer size: 31\u001b[0m\n",
            "\u001b[0mMaximum g-mer size: 31\u001b[0m\n",
            "-- Performing Test CMAKE_HAVE_LIBC_PTHREAD\n",
            "-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed\n",
            "-- Looking for pthread_create in pthreads\n",
            "-- Looking for pthread_create in pthreads - not found\n",
            "-- Looking for pthread_create in pthread\n",
            "-- Looking for pthread_create in pthread - found\n",
            "-- Found Threads: TRUE  \n",
            "-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version \"1.2.11\") \n",
            "-- Configuring done\n",
            "-- Generating done\n",
            "-- Build files have been written to: /content/kallisto-D/ext/bifrost/build\n",
            "[ 23%] \u001b[34m\u001b[1mPerforming build step for 'bifrost'\u001b[0m\n",
            "[  3%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/Bifrost.cpp.o\u001b[0m\n",
            "[  6%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/BitContainer.cpp.o\u001b[0m\n",
            "[  9%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/BlockedBloomFilter.cpp.o\u001b[0m\n",
            "[ 12%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/ColorSet.cpp.o\u001b[0m\n",
            "[ 15%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/CompressedCoverage.cpp.o\u001b[0m\n",
            "[ 18%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/CompressedSequence.cpp.o\u001b[0m\n",
            "[ 21%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/FASTX_Parser.cpp.o\u001b[0m\n",
            "[ 25%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/GFA_Parser.cpp.o\u001b[0m\n",
            "[ 28%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/Kmer.cpp.o\u001b[0m\n",
            "[ 31%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/KmerIterator.cpp.o\u001b[0m\n",
            "[ 34%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/MinimizerIndex.cpp.o\u001b[0m\n",
            "[ 37%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/TinyBitmap.cpp.o\u001b[0m\n",
            "[ 40%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_static.dir/UnitigMap.cpp.o\u001b[0m\n",
            "[ 43%] \u001b[32mBuilding C object src/CMakeFiles/bifrost_static.dir/roaring.c.o\u001b[0m\n",
            "[ 46%] \u001b[32m\u001b[1mLinking CXX static library libbifrost.a\u001b[0m\n",
            "[ 46%] Built target bifrost_static\n",
            "[ 50%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/Bifrost.cpp.o\u001b[0m\n",
            "[ 53%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/BitContainer.cpp.o\u001b[0m\n",
            "[ 56%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/BlockedBloomFilter.cpp.o\u001b[0m\n",
            "[ 59%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/ColorSet.cpp.o\u001b[0m\n",
            "[ 62%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/CompressedCoverage.cpp.o\u001b[0m\n",
            "[ 65%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/CompressedSequence.cpp.o\u001b[0m\n",
            "[ 68%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/FASTX_Parser.cpp.o\u001b[0m\n",
            "[ 71%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/GFA_Parser.cpp.o\u001b[0m\n",
            "[ 75%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/Kmer.cpp.o\u001b[0m\n",
            "[ 78%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/KmerIterator.cpp.o\u001b[0m\n",
            "[ 81%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/MinimizerIndex.cpp.o\u001b[0m\n",
            "[ 84%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/TinyBitmap.cpp.o\u001b[0m\n",
            "[ 87%] \u001b[32mBuilding CXX object src/CMakeFiles/bifrost_dynamic.dir/UnitigMap.cpp.o\u001b[0m\n",
            "[ 90%] \u001b[32mBuilding C object src/CMakeFiles/bifrost_dynamic.dir/roaring.c.o\u001b[0m\n",
            "[ 93%] \u001b[32m\u001b[1mLinking CXX shared library libbifrost.so\u001b[0m\n",
            "[ 93%] Built target bifrost_dynamic\n",
            "[ 96%] \u001b[32mBuilding CXX object src/CMakeFiles/Bifrost.dir/Bifrost.cpp.o\u001b[0m\n",
            "[100%] \u001b[32m\u001b[1mLinking CXX executable Bifrost\u001b[0m\n",
            "[100%] Built target Bifrost\n",
            "[ 26%] \u001b[34m\u001b[1mNo install step for 'bifrost'\u001b[0m\n",
            "[ 30%] \u001b[34m\u001b[1mCompleted 'bifrost'\u001b[0m\n",
            "[ 30%] Built target bifrost\n",
            "[ 34%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/BUSData.cpp.o\u001b[0m\n",
            "[ 38%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/BUSTools.cpp.o\u001b[0m\n",
            "[ 42%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/Bootstrap.cpp.o\u001b[0m\n",
            "[ 46%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/GeneModel.cpp.o\u001b[0m\n",
            "[ 50%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/H5Writer.cpp.o\u001b[0m\n",
            "[ 53%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/KmerIndex.cpp.o\u001b[0m\n",
            "[ 57%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/MinCollector.cpp.o\u001b[0m\n",
            "[ 61%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/PlaintextWriter.cpp.o\u001b[0m\n",
            "[ 65%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/ProcessReads.cpp.o\u001b[0m\n",
            "[ 69%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/PseudoBam.cpp.o\u001b[0m\n",
            "[ 73%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/common.cpp.o\u001b[0m\n",
            "[ 76%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/h5utils.cpp.o\u001b[0m\n",
            "[ 80%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/hash.cpp.o\u001b[0m\n",
            "[ 84%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/main.cpp.o\u001b[0m\n",
            "[ 88%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto_core.dir/weights.cpp.o\u001b[0m\n",
            "[ 92%] \u001b[32m\u001b[1mLinking CXX static library libkallisto_core.a\u001b[0m\n",
            "[ 92%] Built target kallisto_core\n",
            "[ 96%] \u001b[32mBuilding CXX object src/CMakeFiles/kallisto.dir/main.cpp.o\u001b[0m\n",
            "[100%] \u001b[32m\u001b[1mLinking CXX executable kallisto\u001b[0m\n",
            "[100%] Built target kallisto\n",
            "[ 30%] Built target bifrost\n",
            "[ 92%] Built target kallisto_core\n",
            "[100%] Built target kallisto\n",
            "\u001b[36mInstall the project...\u001b[0m\n",
            "-- Install configuration: \"Release\"\n",
            "-- Installing: /usr/local/bin/kallisto\n",
            "Cloning into 'bustools'...\n",
            "remote: Enumerating objects: 3444, done.\u001b[K\n",
            "remote: Counting objects: 100% (656/656), done.\u001b[K\n",
            "remote: Compressing objects: 100% (242/242), done.\u001b[K\n",
            "remote: Total 3444 (delta 491), reused 487 (delta 412), pack-reused 2788\u001b[K\n",
            "Receiving objects: 100% (3444/3444), 3.37 MiB | 7.75 MiB/s, done.\n",
            "Resolving deltas: 100% (1370/1370), done.\n",
            "Branch 'dlist' set up to track remote branch 'dlist' from 'origin'.\n",
            "Switched to a new branch 'dlist'\n",
            "-- The C compiler identification is GNU 9.4.0\n",
            "-- The CXX compiler identification is GNU 9.4.0\n",
            "-- Detecting C compiler ABI info\n",
            "-- Detecting C compiler ABI info - done\n",
            "-- Check for working C compiler: /usr/bin/cc - skipped\n",
            "-- Detecting C compile features\n",
            "-- Detecting C compile features - done\n",
            "-- Detecting CXX compiler ABI info\n",
            "-- Detecting CXX compiler ABI info - done\n",
            "-- Check for working CXX compiler: /usr/bin/c++ - skipped\n",
            "-- Detecting CXX compile features\n",
            "-- Detecting CXX compile features - done\n",
            "\u001b[0mrelease mode\u001b[0m\n",
            "-- Performing Test CMAKE_HAVE_LIBC_PTHREAD\n",
            "-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed\n",
            "-- Looking for pthread_create in pthreads\n",
            "-- Looking for pthread_create in pthreads - not found\n",
            "-- Looking for pthread_create in pthread\n",
            "-- Looking for pthread_create in pthread - found\n",
            "-- Found Threads: TRUE  \n",
            "-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version \"1.2.11\") \n",
            "-- Configuring done\n",
            "-- Generating done\n",
            "-- Build files have been written to: /content/bustools/build\n",
            "[  3%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/BUSData.cpp.o\u001b[0m\n",
            "[  7%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/Common.cpp.o\u001b[0m\n",
            "[ 11%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_capture.cpp.o\u001b[0m\n",
            "[ 15%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_clusterhist.cpp.o\u001b[0m\n",
            "[ 19%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_collapse.cpp.o\u001b[0m\n",
            "[ 23%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_compress.cpp.o\u001b[0m\n",
            "[ 26%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_correct.cpp.o\u001b[0m\n",
            "[ 30%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_count.cpp.o\u001b[0m\n",
            "[ 34%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_decompress.cpp.o\u001b[0m\n",
            "[ 38%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_extract.cpp.o\u001b[0m\n",
            "[ 42%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_inspect.cpp.o\u001b[0m\n",
            "[ 46%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_linker.cpp.o\u001b[0m\n",
            "[ 50%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_main.cpp.o\u001b[0m\n",
            "[ 53%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_mash.cpp.o\u001b[0m\n",
            "[ 57%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_merge.cpp.o\u001b[0m\n",
            "[ 61%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_predict.cpp.o\u001b[0m\n",
            "[ 65%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_project.cpp.o\u001b[0m\n",
            "[ 69%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_sort.cpp.o\u001b[0m\n",
            "[ 73%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_text.cpp.o\u001b[0m\n",
            "[ 76%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_umicorrect.cpp.o\u001b[0m\n",
            "[ 80%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_whitelist.cpp.o\u001b[0m\n",
            "[ 84%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/hash.cpp.o\u001b[0m\n",
            "[ 88%] \u001b[32mBuilding C object src/CMakeFiles/bustools_core.dir/roaring.c.o\u001b[0m\n",
            "[ 92%] \u001b[32m\u001b[1mLinking CXX static library libbustools_core.a\u001b[0m\n",
            "[ 92%] Built target bustools_core\n",
            "[ 96%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools.dir/bustools_main.cpp.o\u001b[0m\n",
            "[100%] \u001b[32m\u001b[1mLinking CXX executable bustools\u001b[0m\n",
            "[100%] Built target bustools\n",
            "[ 92%] Built target bustools_core\n",
            "[100%] Built target bustools\n",
            "\u001b[36mInstall the project...\u001b[0m\n",
            "-- Install configuration: \"Release\"\n",
            "-- Installing: /usr/local/bin/bustools\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Download reference and scRNAseq data"
      ],
      "metadata": {
        "id": "9diQpPTDA2BP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Download host transcriptome to mask host sequences in the virus index\n",
        "!gget ref -d -w dna macaca_mulatta"
      ],
      "metadata": {
        "id": "VtqWrGi6UDF3",
        "outputId": "e9659fec-035f-4ad1-85af-a9a0aae5ce6c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mon May 29 22:43:14 2023 INFO Fetching reference information for macaca_mulatta from Ensembl release: 109.\n",
            "{\n",
            "    \"macaca_mulatta\": {\n",
            "        \"genome_dna\": {\n",
            "            \"ftp\": \"http://ftp.ensembl.org/pub/release-109/fasta/macaca_mulatta/dna/Macaca_mulatta.Mmul_10.dna.toplevel.fa.gz\",\n",
            "            \"ensembl_release\": 109,\n",
            "            \"release_date\": \"2022-12-13\",\n",
            "            \"release_time\": \"00:13\",\n",
            "            \"bytes\": \"832M\"\n",
            "        }\n",
            "    }\n",
            "}\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100  832M  100  832M    0     0   451k      0  0:31:26  0:31:26 --:--:--  452k\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "For this tutorial, we are downloading just 1 of 106 samples available from this paper:"
      ],
      "metadata": {
        "id": "-1-V1JPZVrEt"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Get download links for scRNAseq data using ffq\n",
        "!ffq --ftp SRR12698503"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_gx8VcSrUH_N",
        "outputId": "dec0a6f1-0023-43cf-bd5d-8163a7935fb0"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-05-29 23:14:41,671]    INFO Parsing run SRR12698503\n",
            "[\n",
            "    {\n",
            "        \"accession\": \"SRR12698503\",\n",
            "        \"filename\": \"SRR12698503_1.fastq.gz\",\n",
            "        \"filetype\": \"fastq\",\n",
            "        \"filesize\": 4627785755,\n",
            "        \"filenumber\": 1,\n",
            "        \"md5\": \"5d5b86428390f5954f80db00670cbfd8\",\n",
            "        \"urltype\": \"ftp\",\n",
            "        \"url\": \"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR126/003/SRR12698503/SRR12698503_1.fastq.gz\"\n",
            "    },\n",
            "    {\n",
            "        \"accession\": \"SRR12698503\",\n",
            "        \"filename\": \"SRR12698503_2.fastq.gz\",\n",
            "        \"filetype\": \"fastq\",\n",
            "        \"filesize\": 12900068281,\n",
            "        \"filenumber\": 2,\n",
            "        \"md5\": \"a64337dd64476c7f18e3a4658448e919\",\n",
            "        \"urltype\": \"ftp\",\n",
            "        \"url\": \"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR126/003/SRR12698503/SRR12698503_2.fastq.gz\"\n",
            "    }\n",
            "]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download scRNAseq data\n",
        "!curl -O ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR126/003/SRR12698503/SRR12698503_1.fastq.gz\n",
        "!curl -O ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR126/003/SRR12698503/SRR12698503_2.fastq.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_nOGn8IrAMB8",
        "outputId": "ce26d110-cf29-4d51-9d10-b4c980378647"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 4413M  100 4413M    0     0  20.5M      0  0:03:34  0:03:34 --:--:-- 20.5M\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 12.0G  100 12.0G    0     0  21.1M      0  0:09:40  0:09:40 --:--:-- 20.6M\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download fasta (and annotations) containing PalmDB amino acid reference sequences\n",
        "!curl -O https://raw.githubusercontent.com/lauraluebbert/kallistoXpalmdb/main/palmdb_clustered.fasta\n",
        "!curl -O https://raw.githubusercontent.com/lauraluebbert/kallistoXpalmdb/main/palmdb_clustered.gtf\n",
        "\n",
        "# Download virus ID to taxonomy mapping\n",
        "!curl -O https://raw.githubusercontent.com/lauraluebbert/kallistoXpalmdb/main/palmID_to_taxonomy_mapping_clustered.tsv"
      ],
      "metadata": {
        "id": "5iLX5dsw4-TU",
        "outputId": "dff45f12-41fd-480b-c613-dcaa39b87d8a",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100  819k  100  819k    0     0  1978k      0 --:--:-- --:--:-- --:--:-- 1973k\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 8051k  100 8051k    0     0  14.0M      0 --:--:-- --:--:-- --:--:-- 14.0M\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 14.5M  100 14.5M    0     0  16.5M      0 --:--:-- --:--:-- --:--:-- 16.5M\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Build virus reference index from PalmDB amino acid sequences\n",
        "You can find the kallisto bustools manual and tutorials [here](https://www.kallistobus.tools/).\n",
        "\n",
        "The `--aa` argument tells kb that this is an amino acid reference.  \n",
        "\n",
        "The `--d-list` argument is the path to the **host** transcriptome. These sequences will be masked in the index.\n",
        "\n",
        "Since we installed kallisto from source, we have to tell kb where to find the kallisto binary using the `--kallisto` argument. This will not be neccesary anymore after the official release.\n",
        "\n",
        "(Note: The `--d-list` and `--aa` arguments currently significantly slow down both the `kb ref` and the `kb count` commands. However, we already have some ideas to speed things up before the official release.)"
      ],
      "metadata": {
        "id": "gc_Cfc4r33EG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "!kb ref \\\n",
        "  --aa \\\n",
        "  --d-list Macaca_mulatta.Mmul_10.dna.toplevel.fa.gz \\\n",
        "  -t 2 \\\n",
        "  -i index.idx \\\n",
        "  -g t2g.txt \\\n",
        "  -f1 transcriptome.fa \\\n",
        "  --kallisto /usr/local/bin/kallisto \\\n",
        "  --bustools /usr/local/bin/bustools \\\n",
        "  palmdb_clustered.fasta \\\n",
        "  palmdb_clustered.gtf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WzvFDiRX-NE6",
        "outputId": "ea016e37-6af8-4944-c7ec-41162212c830"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-05-29 23:28:11,687]    INFO [ref] Preparing palmdb_clustered.fasta, palmdb_clustered.gtf\n",
            "[2023-05-29 23:28:12,587]    INFO [ref] Splitting genome palmdb_clustered.fasta into cDNA at /content/tmp/tmpfucqapaz\n",
            "[2023-05-29 23:28:40,080]    INFO [ref] Concatenating 1 cDNAs to transcriptome.fa\n",
            "[2023-05-29 23:28:40,088]    INFO [ref] Creating transcript-to-gene mapping at t2g.txt\n",
            "[2023-05-29 23:28:40,096]    INFO [ref] Indexing transcriptome.fa to index.idx\n",
            "CPU times: user 11.4 s, sys: 1.76 s, total: 13.1 s\n",
            "Wall time: 28min 26s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Align sequencing data and generate virus count matrix\n",
        "The `-x` techology tells kb where to find the barcode and UMI. Here, I added the x string for Seqwell. To view named technologies for `-x`, run `!kb --list`."
      ],
      "metadata": {
        "id": "XhaG2fdy35Jh"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!kb --list"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nM_4_4e5U3li",
        "outputId": "ff237e0c-3718-4f10-e379-d59302feafb9"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "List of supported single-cell technologies\n",
            "\n",
            "Positions syntax: `input file index, start position, end position`\n",
            "When start & end positions are None, refers to the entire file\n",
            "Custom technologies may be defined by providing a kallisto-supported technology string\n",
            "(see https://pachterlab.github.io/kallisto/manual)\n",
            "\n",
            "name            description                            whitelist    barcode                    umi        cDNA                       \n",
            "------------    -----------------------------------    ---------    -----------------------    -------    -----------------------    \n",
            "10XV1           10x version 1                          yes          0,0,14                     1,0,10     2,None,None                \n",
            "10XV2           10x version 2                          yes          0,0,16                     0,16,26    1,None,None                \n",
            "10XV3           10x version 3                          yes          0,0,16                     0,16,28    1,None,None                \n",
            "10XV3_ULTIMA    10x version 3 sequenced with Ultima    yes          0,22,38                    0,38,50    0,62,None                  \n",
            "BDWTA           BD Rhapsody                            yes          0,0,9 0,21,30 0,43,52      0,52,60    1,None,None                \n",
            "BULK            Bulk (single or paired)                                                                   0,None,None 1,None,None    \n",
            "CELSEQ          CEL-Seq                                             0,0,8                      0,8,12     1,None,None                \n",
            "CELSEQ2         CEL-SEQ version 2                                   0,6,12                     0,0,6      1,None,None                \n",
            "DROPSEQ         DropSeq                                             0,0,12                     0,12,20    1,None,None                \n",
            "INDROPSV1       inDrops version 1                                   0,0,11 0,30,38             0,42,48    1,None,None                \n",
            "INDROPSV2       inDrops version 2                                   1,0,11 1,30,38             1,42,48    0,None,None                \n",
            "INDROPSV3       inDrops version 3                      yes          0,0,8 1,0,8                1,8,14     2,None,None                \n",
            "SCRUBSEQ        SCRB-Seq                                            0,0,6                      0,6,16     1,None,None                \n",
            "SMARTSEQ2       Smart-seq2  (single or paired)                                                            0,None,None 1,None,None    \n",
            "SMARTSEQ3       Smart-seq3                                                                     0,11,19    0,11,None 1,None,None      \n",
            "SPLIT-SEQ       SPLiT-seq                                           1,10,18 1,48,56 1,78,86    1,0,10     0,None,None                \n",
            "SURECELL        SureCell for ddSEQ                                  0,0,6 0,21,27 0,42,48      0,51,59    1,None,None                \n",
            "Visium          10x Visium                             yes          0,0,16                     0,16,28    1,None,None                \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# We will only use the first 100,000,000 reads for this demo\n",
        "!zcat SRR12698503_1.fastq.gz | head -400000000 > SRR12698503_1_short.fastq\n",
        "!zcat SRR12698503_2.fastq.gz | head -400000000 > SRR12698503_2_short.fastq"
      ],
      "metadata": {
        "id": "kWlJCCCCigJO"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "!kb count \\\n",
        "  --verbose \\\n",
        "  --aa \\\n",
        "  --h5ad \\\n",
        "  -t 2 \\\n",
        "  -i index.idx \\\n",
        "  -g t2g.txt \\\n",
        "  -x 0,0,12:0,12,20:1,0,0 \\\n",
        "  -o kb_results \\\n",
        "  --kallisto /usr/local/bin/kallisto \\\n",
        "  --bustools /usr/local/bin/bustools \\\n",
        "  SRR12698503_1_short.fastq \\\n",
        "  SRR12698503_2_short.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fYgxpSf2Rc0-",
        "outputId": "ef6d7157-13f7-4f8a-b6e3-179af747a793"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-05-30 00:02:02,318]   DEBUG [main] Printing verbose output\n",
            "[2023-05-30 00:02:04,523]   DEBUG [main] kallisto binary located at /usr/local/bin/kallisto\n",
            "[2023-05-30 00:02:04,524]   DEBUG [main] bustools binary located at /usr/local/bin/bustools\n",
            "[2023-05-30 00:02:04,524]   DEBUG [main] Creating `kb_results/tmp` directory\n",
            "[2023-05-30 00:02:04,527]   DEBUG [main] Namespace(list=False, command='count', tmp=None, keep_tmp=False, verbose=True, i='index.idx', g='t2g.txt', x='0,0,12:0,12,20:1,0,0', o='kb_results', w=None, r=None, t=2, m='2G', strand=None, inleaved=False, genomebam=False, aa=True, gtf=None, chromosomes=None, workflow='standard', em=False, mm=False, tcc=False, filter=None, filter_threshold=None, c1=None, c2=None, overwrite=False, dry_run=False, batch_barcodes=False, loom=False, h5ad=True, loom_names='barcode,target_name', sum='none', cellranger=False, gene_names=False, report=False, no_inspect=False, kallisto='/usr/local/bin/kallisto', bustools='/usr/local/bin/bustools', no_validate=False, parity=None, fragment_l=None, fragment_s=None, fastqs=['SRR12698503_1_short.fastq', 'SRR12698503_2_short.fastq'])\n",
            "[2023-05-30 00:02:07,215]    INFO [count] Using index index.idx to generate BUS file to kb_results from\n",
            "[2023-05-30 00:02:07,215]    INFO [count]         SRR12698503_1_short.fastq\n",
            "[2023-05-30 00:02:07,215]    INFO [count]         SRR12698503_2_short.fastq\n",
            "[2023-05-30 00:02:07,215]   DEBUG [count] kallisto bus -i index.idx -o kb_results -x 0,0,12:0,12,20:1,0,0 -t 2 --aa SRR12698503_1_short.fastq SRR12698503_2_short.fastq\n",
            "[2023-05-30 00:02:07,316]   DEBUG [count] \n",
            "[2023-05-30 00:02:07,317]   DEBUG [count] [bus] Note: Strand option was not specified; setting it to --unstranded for specified technology\n",
            "[2023-05-30 00:02:07,417]   DEBUG [count] [index] k-mer length: 31\n",
            "[2023-05-30 00:02:07,517]   DEBUG [count] [index] number of targets: 6,518\n",
            "[2023-05-30 00:02:07,518]   DEBUG [count] [index] number of k-mers: 2,075,208\n",
            "[2023-05-30 00:02:07,518]   DEBUG [count] [index] number of distinguishing flanking k-mers: 138\n",
            "[2023-05-30 00:02:07,518]   DEBUG [count] [quant] will process sample 1: SRR12698503_1_short.fastq\n",
            "[2023-05-30 00:02:07,518]   DEBUG [count] SRR12698503_2_short.fastq\n",
            "[2023-05-30 00:04:02,985]   DEBUG [count] [quant] finding pseudoalignments for the reads ...\n",
            "[2023-05-30 00:06:03,814]   DEBUG [count] [progress] 1M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:07:58,090]   DEBUG [count] [progress] 2M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:09:52,112]   DEBUG [count] [progress] 3M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:11:46,715]   DEBUG [count] [progress] 4M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:13:40,106]   DEBUG [count] [progress] 5M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:15:34,012]   DEBUG [count] [progress] 6M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:17:36,037]   DEBUG [count] [progress] 7M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:19:29,705]   DEBUG [count] [progress] 8M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:21:23,121]   DEBUG [count] [progress] 9M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:23:18,074]   DEBUG [count] [progress] 10M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:25:11,981]   DEBUG [count] [progress] 11M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:27:10,633]   DEBUG [count] [progress] 12M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:29:05,861]   DEBUG [count] [progress] 13M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:30:59,470]   DEBUG [count] [progress] 14M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:32:53,114]   DEBUG [count] [progress] 16M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:34:47,000]   DEBUG [count] [progress] 17M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:36:48,514]   DEBUG [count] [progress] 18M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:38:44,238]   DEBUG [count] [progress] 19M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:40:40,066]   DEBUG [count] [progress] 20M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:42:35,600]   DEBUG [count] [progress] 21M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:44:29,607]   DEBUG [count] [progress] 22M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:46:32,822]   DEBUG [count] [progress] 23M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:48:28,994]   DEBUG [count] [progress] 24M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:50:26,355]   DEBUG [count] [progress] 25M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:52:23,097]   DEBUG [count] [progress] 26M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:54:19,748]   DEBUG [count] [progress] 27M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:56:16,889]   DEBUG [count] [progress] 28M reads processed (0.4% mapped)\n",
            "[2023-05-30 00:58:20,467]   DEBUG [count] [progress] 29M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:00:24,941]   DEBUG [count] [progress] 30M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:02:24,849]   DEBUG [count] [progress] 32M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:04:23,502]   DEBUG [count] [progress] 33M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:06:17,200]   DEBUG [count] [progress] 34M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:08:21,075]   DEBUG [count] [progress] 35M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:10:14,454]   DEBUG [count] [progress] 36M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:12:12,368]   DEBUG [count] [progress] 37M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:14:08,270]   DEBUG [count] [progress] 38M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:16:04,453]   DEBUG [count] [progress] 39M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:18:06,336]   DEBUG [count] [progress] 40M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:20:01,132]   DEBUG [count] [progress] 41M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:21:55,183]   DEBUG [count] [progress] 42M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:23:52,745]   DEBUG [count] [progress] 43M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:25:48,231]   DEBUG [count] [progress] 44M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:27:48,270]   DEBUG [count] [progress] 45M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:29:42,748]   DEBUG [count] [progress] 46M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:31:38,812]   DEBUG [count] [progress] 48M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:33:36,061]   DEBUG [count] [progress] 49M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:35:29,927]   DEBUG [count] [progress] 50M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:37:33,608]   DEBUG [count] [progress] 51M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:39:27,704]   DEBUG [count] [progress] 52M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:41:21,119]   DEBUG [count] [progress] 53M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:43:11,593]   DEBUG [count] [progress] 54M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:45:01,788]   DEBUG [count] [progress] 55M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:46:54,078]   DEBUG [count] [progress] 56M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:48:51,211]   DEBUG [count] [progress] 57M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:50:43,076]   DEBUG [count] [progress] 58M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:52:33,796]   DEBUG [count] [progress] 59M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:54:23,689]   DEBUG [count] [progress] 60M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:56:14,821]   DEBUG [count] [progress] 61M reads processed (0.4% mapped)\n",
            "[2023-05-30 01:58:11,907]   DEBUG [count] [progress] 62M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:00:04,082]   DEBUG [count] [progress] 64M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:01:55,622]   DEBUG [count] [progress] 65M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:03:46,476]   DEBUG [count] [progress] 66M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:05:37,280]   DEBUG [count] [progress] 67M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:07:28,199]   DEBUG [count] [progress] 68M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:09:27,326]   DEBUG [count] [progress] 69M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:11:17,670]   DEBUG [count] [progress] 70M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:13:10,632]   DEBUG [count] [progress] 71M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:14:59,586]   DEBUG [count] [progress] 72M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:16:54,081]   DEBUG [count] [progress] 73M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:18:50,718]   DEBUG [count] [progress] 74M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:20:42,705]   DEBUG [count] [progress] 75M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:22:32,025]   DEBUG [count] [progress] 76M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:24:24,191]   DEBUG [count] [progress] 77M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:26:15,472]   DEBUG [count] [progress] 79M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:28:08,495]   DEBUG [count] [progress] 80M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:30:05,628]   DEBUG [count] [progress] 81M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:31:54,707]   DEBUG [count] [progress] 82M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:33:47,386]   DEBUG [count] [progress] 83M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:35:36,816]   DEBUG [count] [progress] 84M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:37:29,219]   DEBUG [count] [progress] 85M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:39:26,056]   DEBUG [count] [progress] 86M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:41:18,496]   DEBUG [count] [progress] 87M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:43:08,220]   DEBUG [count] [progress] 88M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:45:00,149]   DEBUG [count] [progress] 89M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:46:50,314]   DEBUG [count] [progress] 90M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:48:48,542]   DEBUG [count] [progress] 91M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:50:37,333]   DEBUG [count] [progress] 92M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:52:28,387]   DEBUG [count] [progress] 93M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:54:18,649]   DEBUG [count] [progress] 95M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:56:09,691]   DEBUG [count] [progress] 96M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:58:00,617]   DEBUG [count] [progress] 97M reads processed (0.4% mapped)\n",
            "[2023-05-30 02:59:58,346]   DEBUG [count] [progress] 98M reads processed (0.4% mapped)\n",
            "[2023-05-30 03:01:06,675]   DEBUG [count] [progress] 99M reads processed (0.4% mapped)              done\n",
            "[2023-05-30 03:01:06,675]   DEBUG [count] [quant] processed 100,000,000 reads, 363,451 reads pseudoaligned\n",
            "[2023-05-30 03:01:06,675]   DEBUG [count] \n",
            "[2023-05-30 03:01:07,887]   DEBUG [count] kb_results/output.bus passed validation\n",
            "[2023-05-30 03:01:07,887]    INFO [count] Sorting BUS file kb_results/output.bus to kb_results/tmp/output.s.bus\n",
            "[2023-05-30 03:01:07,888]   DEBUG [count] bustools sort -o kb_results/tmp/output.s.bus -T kb_results/tmp -t 2 -m 2G kb_results/output.bus\n",
            "[2023-05-30 03:01:11,200]   DEBUG [count] partition time: 0.004016s\n",
            "[2023-05-30 03:01:11,301]   DEBUG [count] all fits in buffer\n",
            "[2023-05-30 03:01:12,502]   DEBUG [count] Read in 363451 BUS records\n",
            "[2023-05-30 03:01:12,502]   DEBUG [count] reading time 0.003803s\n",
            "[2023-05-30 03:01:12,502]   DEBUG [count] sorting time 0.062201s\n",
            "[2023-05-30 03:01:12,502]   DEBUG [count] writing time 0.017463s\n",
            "[2023-05-30 03:01:13,605]   DEBUG [count] kb_results/tmp/output.s.bus passed validation\n",
            "[2023-05-30 03:01:13,605]    INFO [count] Whitelist not provided\n",
            "[2023-05-30 03:01:13,606]    INFO [count] Generating whitelist kb_results/whitelist.txt from BUS file kb_results/tmp/output.s.bus\n",
            "[2023-05-30 03:01:13,606]   DEBUG [count] bustools whitelist -o kb_results/whitelist.txt kb_results/tmp/output.s.bus\n",
            "[2023-05-30 03:01:14,708]   DEBUG [count] Read in 145670 BUS records, wrote 83 barcodes to whitelist with threshold 346\n",
            "[2023-05-30 03:01:14,709]    INFO [count] Inspecting BUS file kb_results/tmp/output.s.bus\n",
            "[2023-05-30 03:01:14,709]   DEBUG [count] bustools inspect -o kb_results/inspect.json -w kb_results/whitelist.txt kb_results/tmp/output.s.bus\n",
            "[2023-05-30 03:01:15,811]    INFO [count] Correcting BUS records in kb_results/tmp/output.s.bus to kb_results/tmp/output.s.c.bus with whitelist kb_results/whitelist.txt\n",
            "[2023-05-30 03:01:15,812]   DEBUG [count] bustools correct -o kb_results/tmp/output.s.c.bus -w kb_results/whitelist.txt kb_results/tmp/output.s.bus\n",
            "[2023-05-30 03:01:16,913]   DEBUG [count] Found 83 barcodes in the on-list\n",
            "[2023-05-30 03:01:16,913]   DEBUG [count] Processed 145670 BUS records\n",
            "[2023-05-30 03:01:16,914]   DEBUG [count] In on-list = 704\n",
            "[2023-05-30 03:01:16,914]   DEBUG [count] Corrected    = 959\n",
            "[2023-05-30 03:01:16,914]   DEBUG [count] Uncorrected  = 144007\n",
            "[2023-05-30 03:01:18,018]   DEBUG [count] kb_results/tmp/output.s.c.bus passed validation\n",
            "[2023-05-30 03:01:18,018]    INFO [count] Sorting BUS file kb_results/tmp/output.s.c.bus to kb_results/output.unfiltered.bus\n",
            "[2023-05-30 03:01:18,018]   DEBUG [count] bustools sort -o kb_results/output.unfiltered.bus -T kb_results/tmp -t 2 -m 2G kb_results/tmp/output.s.c.bus\n",
            "[2023-05-30 03:01:19,222]   DEBUG [count] all fits in buffer\n",
            "[2023-05-30 03:01:20,323]   DEBUG [count] Read in 1663 BUS records\n",
            "[2023-05-30 03:01:20,324]   DEBUG [count] reading time 2.2e-05s\n",
            "[2023-05-30 03:01:20,324]   DEBUG [count] sorting time 0.00015s\n",
            "[2023-05-30 03:01:20,324]   DEBUG [count] writing time 0.000243s\n",
            "[2023-05-30 03:01:21,426]   DEBUG [count] kb_results/output.unfiltered.bus passed validation\n",
            "[2023-05-30 03:01:21,427]    INFO [count] Generating count matrix kb_results/counts_unfiltered/cells_x_genes from BUS file kb_results/output.unfiltered.bus\n",
            "[2023-05-30 03:01:21,427]   DEBUG [count] bustools count -o kb_results/counts_unfiltered/cells_x_genes -g t2g.txt -e kb_results/matrix.ec -t kb_results/transcripts.txt --genecounts --umi-gene kb_results/output.unfiltered.bus\n",
            "[2023-05-30 03:01:22,543]   DEBUG [count] kb_results/counts_unfiltered/cells_x_genes.mtx passed validation\n",
            "[2023-05-30 03:01:22,544]    INFO [count] Reading matrix kb_results/counts_unfiltered/cells_x_genes.mtx\n",
            "[2023-05-30 03:01:22,710]    INFO [count] Writing matrix to h5ad kb_results/counts_unfiltered/adata.h5ad\n",
            "[2023-05-30 03:01:22,791]   DEBUG [main] Removing `kb_results/tmp` directory\n",
            "CPU times: user 1min 3s, sys: 9.16 s, total: 1min 12s\n",
            "Wall time: 2h 59min 26s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "!kb count \\\n",
        "    --aa \\\n",
        "    -t 2 \\\n",
        "    -i index.idx \\\n",
        "    -g t2g.txt \\\n",
        "    --parity single \\\n",
        "    -x default \\\n",
        "    --h5ad \\\n",
        "    --kallisto \"/usr/local/bin/kallisto\" \\\n",
        "    --bustools \"/usr/local/bin/bustools\" \\\n",
        "    -o test \\\n",
        "    SRR12698503_2_short.fastq"
      ],
      "metadata": {
        "id": "9D6KkLBsz8Q_",
        "outputId": "efe025b2-faee-4883-8216-38fb91780256",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-05-30 04:04:27,053]    INFO [count] Using index index.idx to generate BUS file to test from\n",
            "[2023-05-30 04:04:27,053]    INFO [count]         SRR12698503_2_short.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Load results and add metadata for this sample:"
      ],
      "metadata": {
        "id": "NzLKmEhVVzLh"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!curl -O https://raw.githubusercontent.com/lauraluebbert/kallistoXpalmdb/main/example_metadata.txt\n",
        "metadata = pd.read_csv(\"example_metadata.txt\", sep=\",\")\n",
        "\n",
        "# Only keep relevent data\n",
        "metadata[\"mdck_spike_in\"] = metadata[\"mdck_spike_in\"].fillna(False).values\n",
        "metadata = metadata[[\"Run\", \"donor_animal\", \"Experiment\", \"mdck_spike_in\", \"hours_post_innoculation\", \"day_post_infection\"]]\n",
        "metadata.head()"
      ],
      "metadata": {
        "id": "rPZAK0yIVviQ",
        "outputId": "bc8639ef-de4d-4dc9-8552-303b55b3792d",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 134
        }
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100   825  100   825    0     0   3873      0 --:--:-- --:--:-- --:--:--  3873\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "           Run donor_animal  Experiment  mdck_spike_in  \\\n",
              "0  SRR12698503         NHP1  SRX9178101          False   \n",
              "\n",
              "   hours_post_innoculation  day_post_infection  \n",
              "0                        4                 NaN  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-80abeb47-de6a-48a3-b967-57cc11e91e77\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Run</th>\n",
              "      <th>donor_animal</th>\n",
              "      <th>Experiment</th>\n",
              "      <th>mdck_spike_in</th>\n",
              "      <th>hours_post_innoculation</th>\n",
              "      <th>day_post_infection</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-80abeb47-de6a-48a3-b967-57cc11e91e77')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-80abeb47-de6a-48a3-b967-57cc11e91e77 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-80abeb47-de6a-48a3-b967-57cc11e91e77');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Load AnnData object retured by kb-python containing virus counts\n",
        "adata = anndata.read(\"kb_results/counts_unfiltered/adata.h5ad\")\n",
        "adata"
      ],
      "metadata": {
        "id": "ci7bRvDpo5w_",
        "outputId": "0e427201-0bd6-4b11-a758-24567d05fe8d",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "AnnData object with n_obs × n_vars = 83 × 6518"
            ]
          },
          "metadata": {},
          "execution_count": 12
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "adata.X[:, adata.var.index == \"u55137\"].sum()"
      ],
      "metadata": {
        "id": "aNvOCL4Sy9RR",
        "outputId": "5ed940b8-abf9-49a2-c9f7-aba860e0b860",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 38,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0.0"
            ]
          },
          "metadata": {},
          "execution_count": 38
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "adata.X.sum()"
      ],
      "metadata": {
        "id": "biw3kvsFy45w",
        "outputId": "49d36256-ac89-4b27-8ddf-84a16dc48556",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 37,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "835.0"
            ]
          },
          "metadata": {},
          "execution_count": 37
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "adata.obs[\"barcode\"] = adata.obs.index\n",
        "adata.obs[\"sample\"] = \"SRR12698503\"\n",
        "\n",
        "adata.obs = (\n",
        "        adata\n",
        "        .obs.merge(metadata, left_on=\"sample\", right_on=\"Run\", how=\"left\")\n",
        "        .set_index(\"barcode\")\n",
        "    )"
      ],
      "metadata": {
        "id": "Y00adpCbWZkR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Create clean timepoint column\n",
        "# Join day_post_infection and hours_post_innoculation columns\n",
        "adata.obs[\"dpi\"] = adata.obs[\"day_post_infection\"].fillna(adata.obs[\"hours_post_innoculation\"]).astype(int)\n",
        "\n",
        "# Add h/d accession to denote hours/days \n",
        "adata.obs[\"dpi_accessions\"] = np.where(adata.obs[\"hours_post_innoculation\"].isna(), \"d\", \"h\")\n",
        "adata.obs[\"dpi_clean\"] = adata.obs[\"dpi\"].astype(str) + adata.obs[\"dpi_accessions\"].astype(str)"
      ],
      "metadata": {
        "id": "pDY_-j0PXLKp"
      },
      "execution_count": 17,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "adata.obs"
      ],
      "metadata": {
        "id": "wdji7p9OxGXu",
        "outputId": "ea3b8918-77e0-4615-f058-d8da941fdc98",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 455
        }
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                   sample          Run donor_animal  Experiment  \\\n",
              "barcode                                                           \n",
              "AAACTGTCTCTT  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "AAAGGCTGTCTC  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "AACCTGTCTCTT  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "AAGCACTGTCTC  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "AAGCAGCTGTCT  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "...                   ...          ...          ...         ...   \n",
              "TGTCTGTCTCTT  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "TTACCCTGTCTC  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "TTACTGTCTCTT  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "TTCCTGTCTCTT  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "TTGCTGTCTCTT  SRR12698503  SRR12698503         NHP1  SRX9178101   \n",
              "\n",
              "              mdck_spike_in  hours_post_innoculation  day_post_infection  dpi  \\\n",
              "barcode                                                                         \n",
              "AAACTGTCTCTT          False                        4                 NaN    4   \n",
              "AAAGGCTGTCTC          False                        4                 NaN    4   \n",
              "AACCTGTCTCTT          False                        4                 NaN    4   \n",
              "AAGCACTGTCTC          False                        4                 NaN    4   \n",
              "AAGCAGCTGTCT          False                        4                 NaN    4   \n",
              "...                     ...                      ...                 ...  ...   \n",
              "TGTCTGTCTCTT          False                        4                 NaN    4   \n",
              "TTACCCTGTCTC          False                        4                 NaN    4   \n",
              "TTACTGTCTCTT          False                        4                 NaN    4   \n",
              "TTCCTGTCTCTT          False                        4                 NaN    4   \n",
              "TTGCTGTCTCTT          False                        4                 NaN    4   \n",
              "\n",
              "             dpi_accessions dpi_clean  \n",
              "barcode                                \n",
              "AAACTGTCTCTT              h        4h  \n",
              "AAAGGCTGTCTC              h        4h  \n",
              "AACCTGTCTCTT              h        4h  \n",
              "AAGCACTGTCTC              h        4h  \n",
              "AAGCAGCTGTCT              h        4h  \n",
              "...                     ...       ...  \n",
              "TGTCTGTCTCTT              h        4h  \n",
              "TTACCCTGTCTC              h        4h  \n",
              "TTACTGTCTCTT              h        4h  \n",
              "TTCCTGTCTCTT              h        4h  \n",
              "TTGCTGTCTCTT              h        4h  \n",
              "\n",
              "[83 rows x 10 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-f18f181d-e1ae-4158-b3fc-9ec43c7fd019\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample</th>\n",
              "      <th>Run</th>\n",
              "      <th>donor_animal</th>\n",
              "      <th>Experiment</th>\n",
              "      <th>mdck_spike_in</th>\n",
              "      <th>hours_post_innoculation</th>\n",
              "      <th>day_post_infection</th>\n",
              "      <th>dpi</th>\n",
              "      <th>dpi_accessions</th>\n",
              "      <th>dpi_clean</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>barcode</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>AAACTGTCTCTT</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AAAGGCTGTCTC</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AACCTGTCTCTT</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AAGCACTGTCTC</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AAGCAGCTGTCT</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TGTCTGTCTCTT</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TTACCCTGTCTC</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TTACTGTCTCTT</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TTCCTGTCTCTT</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TTGCTGTCTCTT</th>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>SRR12698503</td>\n",
              "      <td>NHP1</td>\n",
              "      <td>SRX9178101</td>\n",
              "      <td>False</td>\n",
              "      <td>4</td>\n",
              "      <td>NaN</td>\n",
              "      <td>4</td>\n",
              "      <td>h</td>\n",
              "      <td>4h</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>83 rows × 10 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-f18f181d-e1ae-4158-b3fc-9ec43c7fd019')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-f18f181d-e1ae-4158-b3fc-9ec43c7fd019 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-f18f181d-e1ae-4158-b3fc-9ec43c7fd019');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Visualize virus counts "
      ],
      "metadata": {
        "id": "2GYJbkkV9piY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Load virus ID to taxonomy mapping\n",
        "tax_df = pd.read_csv(\"palmID_to_taxonomy_mapping_clustered.tsv\", sep='\\t')\n",
        "tax_df.head()"
      ],
      "metadata": {
        "id": "KNKKUyVv9y0m",
        "outputId": "f29844ff-88dd-46bf-c897-2d3024f295d3",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        }
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Label         phylum             class         order            family  \\\n",
              "0  u100164              .                 .             .                 .   \n",
              "1   u10019   Pisuviricota                 .             .                 .   \n",
              "2  u100402   Pisuviricota  Duplopiviricetes  Durnavirales  Picobirnaviridae   \n",
              "3  u100562              .                 .             .                 .   \n",
              "4  u100587  Lenarviricota  Allassoviricetes   Levivirales                 .   \n",
              "\n",
              "            genus             species  \n",
              "0               .                   .  \n",
              "1               .                   .  \n",
              "2  Picobirnavirus  Picobirnavirus sp.  \n",
              "3               .                   .  \n",
              "4               .                   .  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-463c9979-28a8-46b8-977d-7f2e4b52e383\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Label</th>\n",
              "      <th>phylum</th>\n",
              "      <th>class</th>\n",
              "      <th>order</th>\n",
              "      <th>family</th>\n",
              "      <th>genus</th>\n",
              "      <th>species</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>u100164</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>u10019</td>\n",
              "      <td>Pisuviricota</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>u100402</td>\n",
              "      <td>Pisuviricota</td>\n",
              "      <td>Duplopiviricetes</td>\n",
              "      <td>Durnavirales</td>\n",
              "      <td>Picobirnaviridae</td>\n",
              "      <td>Picobirnavirus</td>\n",
              "      <td>Picobirnavirus sp.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>u100562</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>u100587</td>\n",
              "      <td>Lenarviricota</td>\n",
              "      <td>Allassoviricetes</td>\n",
              "      <td>Levivirales</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "      <td>.</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-463c9979-28a8-46b8-977d-7f2e4b52e383')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-463c9979-28a8-46b8-977d-7f2e4b52e383 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-463c9979-28a8-46b8-977d-7f2e4b52e383');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 26
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Show Ebolavirus species\n",
        "tax_df[tax_df[\"species\"].str.contains(\"ebolavirus\")]"
      ],
      "metadata": {
        "id": "zk4piHyOIC2F",
        "outputId": "286d49ab-827c-40a2-d834-78d759acf463",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 112
        }
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       Label           phylum           class            order  \\\n",
              "2051  u20513  Negarnaviricota  Monjiviricetes  Mononegavirales   \n",
              "5929  u55137  Negarnaviricota  Monjiviricetes  Mononegavirales   \n",
              "\n",
              "                         family  \\\n",
              "2051  Pneumoviridae_Filoviridae   \n",
              "5929                Filoviridae   \n",
              "\n",
              "                                                  genus  \\\n",
              "2051  Orthopneumovirus_Metapneumovirus_Marburgvirus_...   \n",
              "5929                                         Ebolavirus   \n",
              "\n",
              "                                                species  \n",
              "2051  Human orthopneumovirus_Bovine orthopneumovirus...  \n",
              "5929  Zaire ebolavirus_Sudan ebolavirus_Tai Forest e...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-17dc1800-a255-4033-9d81-4e574a943341\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Label</th>\n",
              "      <th>phylum</th>\n",
              "      <th>class</th>\n",
              "      <th>order</th>\n",
              "      <th>family</th>\n",
              "      <th>genus</th>\n",
              "      <th>species</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2051</th>\n",
              "      <td>u20513</td>\n",
              "      <td>Negarnaviricota</td>\n",
              "      <td>Monjiviricetes</td>\n",
              "      <td>Mononegavirales</td>\n",
              "      <td>Pneumoviridae_Filoviridae</td>\n",
              "      <td>Orthopneumovirus_Metapneumovirus_Marburgvirus_...</td>\n",
              "      <td>Human orthopneumovirus_Bovine orthopneumovirus...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5929</th>\n",
              "      <td>u55137</td>\n",
              "      <td>Negarnaviricota</td>\n",
              "      <td>Monjiviricetes</td>\n",
              "      <td>Mononegavirales</td>\n",
              "      <td>Filoviridae</td>\n",
              "      <td>Ebolavirus</td>\n",
              "      <td>Zaire ebolavirus_Sudan ebolavirus_Tai Forest e...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-17dc1800-a255-4033-9d81-4e574a943341')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-17dc1800-a255-4033-9d81-4e574a943341 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-17dc1800-a255-4033-9d81-4e574a943341');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 28
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "adata.X[:, adata.var.index == \"u55137\"].sum()"
      ],
      "metadata": {
        "id": "PcA_FIYyyUo8",
        "outputId": "fcd309f9-bd36-4b00-d691-62e6c6a6bdc7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0.0"
            ]
          },
          "metadata": {},
          "execution_count": 36
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Print total target virus counts across all samples\n",
        "target_ids = tax_df[tax_df[\"species\"].str.contains(\"ebolavirus\")][\"Label\"].values\n",
        "for id_ in target_ids:\n",
        "    counts = adata.X[:, adata.var.index == id_].sum()\n",
        "    if counts > 0:\n",
        "        print(id_, \": \", counts)"
      ],
      "metadata": {
        "id": "cOVUb4MURGhv"
      },
      "execution_count": 31,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "fig, ax = plt.subplots(figsize=(10, 5))\n",
        "fontsize = 14\n",
        "\n",
        "samples = adata.obs[\"dpi_clean\"]\n",
        "labels = samples\n",
        "virus_id = \"u55137\"\n",
        "\n",
        "cellnum = []\n",
        "counts = []\n",
        "for sample in samples:\n",
        "    count = adata.X[adata.obs[\"dpi_clean\"] == sample, adata.var.index == virus_id].sum()\n",
        "    counts.append(count)\n",
        "    if count > 0:\n",
        "        cellnum.append(len(nd(adata.X[adata.obs[\"dpi_clean\"] == sample, adata.var.index == virus_id])))\n",
        "    else:\n",
        "        cellnum.append(0)\n",
        "\n",
        "fontsize = 14\n",
        "width = 0.4\n",
        "x = np.arange(len(labels))\n",
        "\n",
        "ax.bar(x-width/2, counts, width=width, label=f\"UMI counts\", color=\"#d42526\", edgecolor=\"black\")\n",
        "ax.bar(x+width/2, cellnum, width=width, label=f\"EBOV+ cells\", color=\"#d42526\", alpha=0.5, edgecolor=\"black\")\n",
        "\n",
        "ax.set_yscale(\"log\")\n",
        "ax.set_ylabel(\"Total count\", fontsize=fontsize)\n",
        "ax.set_xlabel(\"Timepoint\", fontsize=fontsize)\n",
        "ax.legend(loc=\"upper left\", fontsize=fontsize)\n",
        "\n",
        "ax.set_xticks(x, samples, rotation=0, ha=\"center\")\n",
        "\n",
        "ax.tick_params(axis=\"both\", labelsize=fontsize)\n",
        "ax.set_title(f\"Zaire Ebolavirus ({virus_id}) RdRP RNA molecule and cell counts\", fontsize=fontsize+2)\n",
        "\n",
        "ax.grid(True, which=\"both\", color=\"lightgray\")\n",
        "ax.set_axisbelow(True)\n",
        "\n",
        "fig.show()"
      ],
      "metadata": {
        "id": "W8HiPFQFdDfs",
        "outputId": "0308bb35-69c7-4a18-bc88-655e947a2a22",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 823
        }
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ValueError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-22-070b1e0df167>\u001b[0m in \u001b[0;36m<cell line: 10>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mcounts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0msample\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msamples\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m     \u001b[0mcount\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"dpi_clean\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0msample\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mvirus_id\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m     \u001b[0mcounts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcount\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/scipy/sparse/_index.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m     87\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     88\u001b[0m         \u001b[0;31m# The only remaining case is inner (fancy) indexing\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 89\u001b[0;31m         \u001b[0mrow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_broadcast_arrays\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     90\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     91\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'number of row and column indices differ'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/scipy/sparse/_index.py\u001b[0m in \u001b[0;36m_broadcast_arrays\u001b[0;34m(a, b)\u001b[0m\n\u001b[1;32m     20\u001b[0m     \u001b[0mthe\u001b[0m \u001b[0mold\u001b[0m \u001b[0mbehavior\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m     \"\"\"\n\u001b[0;32m---> 22\u001b[0;31m     \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbroadcast_arrays\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m     \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriteable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriteable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m     \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriteable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriteable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/numpy/core/overrides.py\u001b[0m in \u001b[0;36mbroadcast_arrays\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/numpy/lib/stride_tricks.py\u001b[0m in \u001b[0;36mbroadcast_arrays\u001b[0;34m(subok, *args)\u001b[0m\n\u001b[1;32m    537\u001b[0m     \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_m\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubok\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msubok\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0m_m\u001b[0m \u001b[0;32min\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    538\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 539\u001b[0;31m     \u001b[0mshape\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_broadcast_shape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    540\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    541\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mshape\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0marray\u001b[0m \u001b[0;32min\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/numpy/lib/stride_tricks.py\u001b[0m in \u001b[0;36m_broadcast_shape\u001b[0;34m(*args)\u001b[0m\n\u001b[1;32m    419\u001b[0m     \u001b[0;31m# use the old-iterator because np.nditer does not handle size 0 arrays\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    420\u001b[0m     \u001b[0;31m# consistently\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 421\u001b[0;31m     \u001b[0mb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbroadcast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m32\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    422\u001b[0m     \u001b[0;31m# unfortunately, it cannot handle 32 or more arguments directly\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    423\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mpos\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m32\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m31\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: shape mismatch: objects cannot be broadcast to a single shape.  Mismatch is between arg 0 with shape (83,) and arg 1 with shape (0,)."
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1000x500 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0UAAAGyCAYAAAArj289AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAe50lEQVR4nO3db2zdVf3A8U/b0VuItAzn2m0WJyigAhturBYkBFNpIhnugaEOsi0LiMgkQKOy8WcV0XUqkCVSXBggPsENCRDCliJUFqLULG5rAnEbwTG2ENptKu0surL2+3tgqL+6Dna7/qE7r1dyH/Rwzv2eSw6DN9/bewuyLMsCAAAgUYVjvQEAAICxJIoAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApOUdRS+99FLMnTs3pk6dGgUFBfH0009/6JqNGzfGF7/4xcjlcvGZz3wmHn300SFsFQAAYPjlHUXd3d0xY8aMaGpqOqr5b7zxRlx++eVx6aWXRltbW9x8881x7bXXxnPPPZf3ZgEAAIZbQZZl2ZAXFxTEU089FfPmzTvinFtvvTXWr18fr776av/YN7/5zXjnnXeiubl5qJcGAAAYFhNG+gKtra1RU1MzYKy2tjZuvvnmI645ePBgHDx4sP/nvr6++Pvf/x4f//jHo6CgYKS2CgAAfMRlWRYHDhyIqVOnRmHh8HxEwohHUXt7e5SXlw8YKy8vj66urvjXv/4VJ5544mFrGhsb46677hrprQEAAOPUnj174pOf/OSwPNeIR9FQLFu2LOrr6/t/7uzsjNNOOy327NkTpaWlY7gzAABgLHV1dUVlZWWcfPLJw/acIx5FFRUV0dHRMWCso6MjSktLB71LFBGRy+Uil8sdNl5aWiqKAACAYf21mhH/nqLq6upoaWkZMPb8889HdXX1SF8aAADgQ+UdRf/85z+jra0t2traIuI/H7nd1tYWu3fvjoj/vPVt4cKF/fOvv/762LlzZ/zgBz+I7du3xwMPPBCPP/543HLLLcPzCgAAAI5B3lH05z//Oc4///w4//zzIyKivr4+zj///Fi+fHlERLz99tv9gRQR8elPfzrWr18fzz//fMyYMSPuvffeeOihh6K2tnaYXgIAAMDQHdP3FI2Wrq6uKCsri87OTr9TBAAACRuJNhjx3ykCAAD4KBNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDShhRFTU1NMX369CgpKYmqqqrYtGnTB85ftWpVnHXWWXHiiSdGZWVl3HLLLfHvf/97SBsGAAAYTnlH0bp166K+vj4aGhpiy5YtMWPGjKitrY29e/cOOv+xxx6LpUuXRkNDQ2zbti0efvjhWLduXdx2223HvHkAAIBjlXcU3XffffGtb30rFi9eHJ///Odj9erVcdJJJ8Ujjzwy6PyXX345Lrroorjqqqti+vTpcdlll8X8+fM/9O4SAADAaMgrinp6emLz5s1RU1Pz3ycoLIyamppobW0ddM2FF14Ymzdv7o+gnTt3xoYNG+JrX/vaEa9z8ODB6OrqGvAAAAAYCRPymbx///7o7e2N8vLyAePl5eWxffv2QddcddVVsX///vjyl78cWZbFoUOH4vrrr//At881NjbGXXfdlc/WAAAAhmTEP31u48aNsWLFinjggQdiy5Yt8eSTT8b69evj7rvvPuKaZcuWRWdnZ/9jz549I71NAAAgUXndKZo0aVIUFRVFR0fHgPGOjo6oqKgYdM2dd94ZCxYsiGuvvTYiIs4999zo7u6O6667Lm6//fYoLDy8y3K5XORyuXy2BgAAMCR53SkqLi6OWbNmRUtLS/9YX19ftLS0RHV19aBr3n333cPCp6ioKCIisizLd78AAADDKq87RRER9fX1sWjRopg9e3bMmTMnVq1aFd3d3bF48eKIiFi4cGFMmzYtGhsbIyJi7ty5cd9998X5558fVVVV8frrr8edd94Zc+fO7Y8jAACAsZJ3FNXV1cW+ffti+fLl0d7eHjNnzozm5ub+D1/YvXv3gDtDd9xxRxQUFMQdd9wRb731VnziE5+IuXPnxk9+8pPhexUAAABDVJCNg/ewdXV1RVlZWXR2dkZpaelYbwcAABgjI9EGI/7pcwAAAB9loggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASNqQoqipqSmmT58eJSUlUVVVFZs2bfrA+e+8804sWbIkpkyZErlcLs4888zYsGHDkDYMAAAwnCbku2DdunVRX18fq1evjqqqqli1alXU1tbGjh07YvLkyYfN7+npia9+9asxefLkeOKJJ2LatGnx5ptvximnnDIc+wcAADgmBVmWZfksqKqqigsuuCDuv//+iIjo6+uLysrKuPHGG2Pp0qWHzV+9enX8/Oc/j+3bt8cJJ5wwpE12dXVFWVlZdHZ2Rmlp6ZCeAwAAGP9Gog3yevtcT09PbN68OWpqav77BIWFUVNTE62trYOueeaZZ6K6ujqWLFkS5eXlcc4558SKFSuit7f3iNc5ePBgdHV1DXgAAACMhLyiaP/+/dHb2xvl5eUDxsvLy6O9vX3QNTt37ownnngient7Y8OGDXHnnXfGvffeGz/+8Y+PeJ3GxsYoKyvrf1RWVuazTQAAgKM24p8+19fXF5MnT44HH3wwZs2aFXV1dXH77bfH6tWrj7hm2bJl0dnZ2f/Ys2fPSG8TAABIVF4ftDBp0qQoKiqKjo6OAeMdHR1RUVEx6JopU6bECSecEEVFRf1jn/vc56K9vT16enqiuLj4sDW5XC5yuVw+WwMAABiSvO4UFRcXx6xZs6KlpaV/rK+vL1paWqK6unrQNRdddFG8/vrr0dfX1z/22muvxZQpUwYNIgAAgNGU99vn6uvrY82aNfHrX/86tm3bFt/5zneiu7s7Fi9eHBERCxcujGXLlvXP/853vhN///vf46abborXXnst1q9fHytWrIglS5YM36sAAAAYory/p6iuri727dsXy5cvj/b29pg5c2Y0Nzf3f/jC7t27o7Dwv61VWVkZzz33XNxyyy1x3nnnxbRp0+Kmm26KW2+9dfheBQAAwBDl/T1FY8H3FAEAABEfge8pAgAAON6IIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaUOKoqamppg+fXqUlJREVVVVbNq06ajWrV27NgoKCmLevHlDuSwAAMCwyzuK1q1bF/X19dHQ0BBbtmyJGTNmRG1tbezdu/cD1+3atSu+973vxcUXXzzkzQIAAAy3vKPovvvui29961uxePHi+PznPx+rV6+Ok046KR555JEjrunt7Y2rr7467rrrrjj99NOPacMAAADDKa8o6unpic2bN0dNTc1/n6CwMGpqaqK1tfWI6370ox/F5MmT45prrjmq6xw8eDC6uroGPAAAAEZCXlG0f//+6O3tjfLy8gHj5eXl0d7ePuiaP/zhD/Hwww/HmjVrjvo6jY2NUVZW1v+orKzMZ5sAAABHbUQ/fe7AgQOxYMGCWLNmTUyaNOmo1y1btiw6Ozv7H3v27BnBXQIAACmbkM/kSZMmRVFRUXR0dAwY7+joiIqKisPm//Wvf41du3bF3Llz+8f6+vr+c+EJE2LHjh1xxhlnHLYul8tFLpfLZ2sAAABDktedouLi4pg1a1a0tLT0j/X19UVLS0tUV1cfNv/ss8+OV155Jdra2vofV1xxRVx66aXR1tbmbXEAAMCYy+tOUUREfX19LFq0KGbPnh1z5syJVatWRXd3dyxevDgiIhYuXBjTpk2LxsbGKCkpiXPOOWfA+lNOOSUi4rBxAACAsZB3FNXV1cW+ffti+fLl0d7eHjNnzozm5ub+D1/YvXt3FBaO6K8qAQAADJuCLMuysd7Eh+nq6oqysrLo7OyM0tLSsd4OAAAwRkaiDdzSAQAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASJooAgAAkiaKAACApIkiAAAgaaIIAABImigCAACSJooAAICkDSmKmpqaYvr06VFSUhJVVVWxadOmI85ds2ZNXHzxxTFx4sSYOHFi1NTUfOB8AACA0ZR3FK1bty7q6+ujoaEhtmzZEjNmzIja2trYu3fvoPM3btwY8+fPjxdffDFaW1ujsrIyLrvssnjrrbeOefMAAADHqiDLsiyfBVVVVXHBBRfE/fffHxERfX19UVlZGTfeeGMsXbr0Q9f39vbGxIkT4/7774+FCxce1TW7urqirKwsOjs7o7S0NJ/tAgAAx5GRaIO87hT19PTE5s2bo6am5r9PUFgYNTU10draelTP8e6778Z7770Xp5566hHnHDx4MLq6ugY8AAAARkJeUbR///7o7e2N8vLyAePl5eXR3t5+VM9x6623xtSpUweE1f9qbGyMsrKy/kdlZWU+2wQAADhqo/rpcytXroy1a9fGU089FSUlJUect2zZsujs7Ox/7NmzZxR3CQAApGRCPpMnTZoURUVF0dHRMWC8o6MjKioqPnDtPffcEytXrowXXnghzjvvvA+cm8vlIpfL5bM1AACAIcnrTlFxcXHMmjUrWlpa+sf6+vqipaUlqqurj7juZz/7Wdx9993R3Nwcs2fPHvpuAQAAhlled4oiIurr62PRokUxe/bsmDNnTqxatSq6u7tj8eLFERGxcOHCmDZtWjQ2NkZExE9/+tNYvnx5PPbYYzF9+vT+3z362Mc+Fh/72MeG8aUAAADkL+8oqquri3379sXy5cujvb09Zs6cGc3Nzf0fvrB79+4oLPzvDahf/vKX0dPTE9/4xjcGPE9DQ0P88Ic/PLbdAwAAHKO8v6doLPieIgAAIOIj8D1FAAAAxxtRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkTRQBAABJE0UAAEDSRBEAAJA0UQQAACRNFAEAAEkTRQAAQNJEEQAAkDRRBAAAJE0UAQAASRNFAABA0kQRAACQNFEEAAAkbUhR1NTUFNOnT4+SkpKoqqqKTZs2feD83/72t3H22WdHSUlJnHvuubFhw4YhbRYAAGC45R1F69ati/r6+mhoaIgtW7bEjBkzora2Nvbu3Tvo/Jdffjnmz58f11xzTWzdujXmzZsX8+bNi1dfffWYNw8AAHCsCrIsy/JZUFVVFRdccEHcf//9ERHR19cXlZWVceONN8bSpUsPm19XVxfd3d3x7LPP9o996UtfipkzZ8bq1auP6ppdXV1RVlYWnZ2dUVpams92AQCA48hItMGEfCb39PTE5s2bY9myZf1jhYWFUVNTE62trYOuaW1tjfr6+gFjtbW18fTTTx/xOgcPHoyDBw/2/9zZ2RkR//kbAAAApOv9Jsjz3s4HyiuK9u/fH729vVFeXj5gvLy8PLZv3z7omvb29kHnt7e3H/E6jY2Ncddddx02XllZmc92AQCA49Tf/va3KCsrG5bnyiuKRsuyZcsG3F1655134lOf+lTs3r172F44DKarqysqKytjz5493qrJiHLWGC3OGqPFWWO0dHZ2xmmnnRannnrqsD1nXlE0adKkKCoqio6OjgHjHR0dUVFRMeiaioqKvOZHRORyucjlcoeNl5WV+YeMUVFaWuqsMSqcNUaLs8ZocdYYLYWFw/ftQnk9U3FxccyaNStaWlr6x/r6+qKlpSWqq6sHXVNdXT1gfkTE888/f8T5AAAAoynvt8/V19fHokWLYvbs2TFnzpxYtWpVdHd3x+LFiyMiYuHChTFt2rRobGyMiIibbropLrnkkrj33nvj8ssvj7Vr18af//znePDBB4f3lQAAAAxB3lFUV1cX+/bti+XLl0d7e3vMnDkzmpub+z9MYffu3QNuZV144YXx2GOPxR133BG33XZbfPazn42nn346zjnnnKO+Zi6Xi4aGhkHfUgfDyVljtDhrjBZnjdHirDFaRuKs5f09RQAAAMeT4fvtJAAAgHFIFAEAAEkTRQAAQNJEEQAAkLSPTBQ1NTXF9OnTo6SkJKqqqmLTpk0fOP+3v/1tnH322VFSUhLnnntubNiwYZR2yniXz1lbs2ZNXHzxxTFx4sSYOHFi1NTUfOjZhPfl++fa+9auXRsFBQUxb968kd0gx418z9o777wTS5YsiSlTpkQul4szzzzTv0c5KvmetVWrVsVZZ50VJ554YlRWVsYtt9wS//73v0dpt4xHL730UsydOzemTp0aBQUF8fTTT3/omo0bN8YXv/jFyOVy8ZnPfCYeffTRvK/7kYiidevWRX19fTQ0NMSWLVtixowZUVtbG3v37h10/ssvvxzz58+Pa665JrZu3Rrz5s2LefPmxauvvjrKO2e8yfesbdy4MebPnx8vvvhitLa2RmVlZVx22WXx1ltvjfLOGW/yPWvv27VrV3zve9+Liy++eJR2yniX71nr6emJr371q7Fr16544oknYseOHbFmzZqYNm3aKO+c8Sbfs/bYY4/F0qVLo6GhIbZt2xYPP/xwrFu3Lm677bZR3jnjSXd3d8yYMSOampqOav4bb7wRl19+eVx66aXR1tYWN998c1x77bXx3HPP5Xfh7CNgzpw52ZIlS/p/7u3tzaZOnZo1NjYOOv/KK6/MLr/88gFjVVVV2be//e0R3SfjX75n7X8dOnQoO/nkk7Nf//rXI7VFjhNDOWuHDh3KLrzwwuyhhx7KFi1alH39618fhZ0y3uV71n75y19mp59+etbT0zNaW+Q4ke9ZW7JkSfaVr3xlwFh9fX120UUXjeg+OX5ERPbUU0994Jwf/OAH2Re+8IUBY3V1dVltbW1e1xrzO0U9PT2xefPmqKmp6R8rLCyMmpqaaG1tHXRNa2vrgPkREbW1tUecDxFDO2v/691334333nsvTj311JHaJseBoZ61H/3oRzF58uS45pprRmObHAeGctaeeeaZqK6ujiVLlkR5eXmcc845sWLFiujt7R2tbTMODeWsXXjhhbF58+b+t9jt3LkzNmzYEF/72tdGZc+kYbi6YMJwbmoo9u/fH729vVFeXj5gvLy8PLZv3z7omvb29kHnt7e3j9g+Gf+Gctb+16233hpTp0497B8++P+Gctb+8Ic/xMMPPxxtbW2jsEOOF0M5azt37ozf//73cfXVV8eGDRvi9ddfjxtuuCHee++9aGhoGI1tMw4N5axdddVVsX///vjyl78cWZbFoUOH4vrrr/f2OYbVkbqgq6sr/vWvf8WJJ554VM8z5neKYLxYuXJlrF27Np566qkoKSkZ6+1wHDlw4EAsWLAg1qxZE5MmTRrr7XCc6+vri8mTJ8eDDz4Ys2bNirq6urj99ttj9erVY701jjMbN26MFStWxAMPPBBbtmyJJ598MtavXx933333WG8NDjPmd4omTZoURUVF0dHRMWC8o6MjKioqBl1TUVGR13yIGNpZe98999wTK1eujBdeeCHOO++8kdwmx4F8z9pf//rX2LVrV8ydO7d/rK+vLyIiJkyYEDt27IgzzjhjZDfNuDSUP9emTJkSJ5xwQhQVFfWPfe5zn4v29vbo6emJ4uLiEd0z49NQztqdd94ZCxYsiGuvvTYiIs4999zo7u6O6667Lm6//fYoLPT/5jl2R+qC0tLSo75LFPERuFNUXFwcs2bNipaWlv6xvr6+aGlpierq6kHXVFdXD5gfEfH8888fcT5EDO2sRUT87Gc/i7vvvjuam5tj9uzZo7FVxrl8z9rZZ58dr7zySrS1tfU/rrjiiv5P0qmsrBzN7TOODOXPtYsuuihef/31/vCOiHjttddiypQpgogjGspZe/fddw8Ln/dj/D+/Qw/Hbti6IL/PgBgZa9euzXK5XPboo49mf/nLX7LrrrsuO+WUU7L29vYsy7JswYIF2dKlS/vn//GPf8wmTJiQ3XPPPdm2bduyhoaG7IQTTsheeeWVsXoJjBP5nrWVK1dmxcXF2RNPPJG9/fbb/Y8DBw6M1UtgnMj3rP0vnz7H0cr3rO3evTs7+eSTs+9+97vZjh07smeffTabPHly9uMf/3isXgLjRL5nraGhITv55JOz3/zmN9nOnTuz3/3ud9kZZ5yRXXnllWP1EhgHDhw4kG3dujXbunVrFhHZfffdl23dujV78803syzLsqVLl2YLFizon79z587spJNOyr7//e9n27Zty5qamrKioqKsubk5r+t+JKIoy7LsF7/4RXbaaadlxcXF2Zw5c7I//elP/X/tkksuyRYtWjRg/uOPP56deeaZWXFxcfaFL3whW79+/SjvmPEqn7P2qU99KouIwx4NDQ2jv3HGnXz/XPv/RBH5yPesvfzyy1lVVVWWy+Wy008/PfvJT36SHTp0aJR3zXiUz1l77733sh/+8IfZGWeckZWUlGSVlZXZDTfckP3jH/8Y/Y0zbrz44ouD/rfX+2dr0aJF2SWXXHLYmpkzZ2bFxcXZ6aefnv3qV7/K+7oFWeb+JQAAkK4x/50iAACAsSSKAACApIkiAAAgaaIIAABImigCAACSJooAAICkiSIAACBpoggAAEiaKAIAAJImigAAgKSJIgAAIGmiCAAASNr/AUOP/hLIsQ49AAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Define virus filtering thresholds for bar and Krona plots:"
      ],
      "metadata": {
        "id": "KK45n2eKRqU8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Minimum count of virus in at least one cell\n",
        "count_threshold = 2\n",
        "# Minimum number of cells the virus was seen in\n",
        "num_cells_threshold = 100"
      ],
      "metadata": {
        "id": "LK7Ny_V0PfYz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "samples = []\n",
        "\n",
        "# Get counts for each virus per sample\n",
        "vir_fractions_df = pd.DataFrame()\n",
        "\n",
        "for idx, virus_id in enumerate(adata.var.index):\n",
        "    if adata.X[:, adata.var.index == virus_id].max() >= count_threshold and len(nd(adata.X[:, adata.var.index == virus_id].todense())) >= num_cells_threshold:\n",
        "        virus_counts = []\n",
        "        \n",
        "        for timepoint in timepoints:\n",
        "            for animal_id in adata.obs[adata.obs[\"dpi_clean\"] == timepoint][\"donor_animal\"].unique():\n",
        "                virus_counts.append(adata.X[(adata.obs[\"donor_animal\"] == animal_id) & (adata.obs[\"dpi_clean\"] == timepoint), adata.var.index == virus_id].sum())\n",
        "                \n",
        "                if idx == 0:\n",
        "                    samples.append(timepoint + \"_\" + animal_id)\n",
        "\n",
        "        utax_label = \"_\".join(phylogeny_data[phylogeny_data[\"Label\"] == virus_id].values[0]).replace(\"_.\", \"\")\n",
        "        vir_fractions_df[utax_label] = virus_counts\n",
        "\n",
        "vir_fractions_df[\"sample\"] = samples\n",
        "\n",
        "# De-fragment dataframe\n",
        "vir_fractions_df = vir_fractions_df.copy()\n",
        "\n",
        "# Set sample as index\n",
        "vir_fractions_df = vir_fractions_df.set_index(\"sample\")\n",
        "\n",
        "# Get total virus reads per sample\n",
        "vir_fractions_df[\"sample_total\"] = vir_fractions_df.sum(axis = 1)\n",
        "\n",
        "# Get fractions for each virus\n",
        "for virus in vir_fractions_df.columns[:-1]:\n",
        "    vir_fractions_df[f\"{virus}_fraction\"] = vir_fractions_df[virus] / vir_fractions_df[\"sample_total\"]\n",
        "    \n",
        "# De-fragment dataframe\n",
        "vir_fractions_df = vir_fractions_df.copy()\n",
        "\n",
        "# Sort by maximum value accross all samples\n",
        "vir_fractions_df = vir_fractions_df[vir_fractions_df.max().sort_values(ascending=False).index]\n",
        "\n",
        "vir_fractions_df.head()"
      ],
      "metadata": {
        "id": "c7M7ZUHdO5j8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "fig, ax = plt.subplots(figsize=(20, 10))\n",
        "fontsize = 14\n",
        "width = 0.75\n",
        "\n",
        "samples = vir_fractions_df.index.values\n",
        "x = np.arange(len(samples))\n",
        "bottom = np.zeros(len(samples))\n",
        "\n",
        "NUM_COLORS = 20\n",
        "color_idx = 0\n",
        "second_hatch_round = False\n",
        "third_hatch_round = False\n",
        "forth_hatch_round = False\n",
        "hatch=\"\"\n",
        "mpl.rcParams['hatch.linewidth'] = 0.5\n",
        "legend_counter = 0\n",
        "\n",
        "for virus_column in vir_fractions_df.columns:\n",
        "    if \"_fraction\" in virus_column:\n",
        "        label_temp = virus_column.split(\"_fraction\")[0]\n",
        "        # Split long labels after n characters\n",
        "        n = 79\n",
        "        label = (\"\\n\".join(label_temp[i:i+n] for i in range(0, len(label_temp), n))) if len(label_temp) > n else label_temp\n",
        "\n",
        "        # Only show labels for first n viruses (else legend will be too long / break plot)\n",
        "        if legend_counter <= 47:\n",
        "            ax.barh(x, vir_fractions_df[virus_column].values, width, label=label, left=bottom, color=plt.cm.tab20(color_idx), hatch=hatch)\n",
        "            legend_counter += 1\n",
        "        else:\n",
        "            ax.barh(x, vir_fractions_df[virus_column].values, width, left=bottom, color=plt.cm.tab20(color_idx), hatch=hatch)\n",
        "        \n",
        "        bottom += vir_fractions_df[virus_column].values\n",
        "\n",
        "        # Change hatch every 20 viruses\n",
        "        if color_idx == NUM_COLORS-1 and not second_hatch_round and not third_hatch_round and not forth_hatch_round:\n",
        "            hatch = \"/\"\n",
        "            second_hatch_round = True\n",
        "            color_idx = 0\n",
        "            continue\n",
        "        elif color_idx == NUM_COLORS-1 and second_hatch_round:\n",
        "            hatch = \"\\\\\"\n",
        "            color_idx = 0\n",
        "            second_hatch_round = False\n",
        "            third_hatch_round = True\n",
        "            continue\n",
        "        elif color_idx == NUM_COLORS-1 and third_hatch_round:\n",
        "            hatch = \"o\"\n",
        "            color_idx = 0\n",
        "            third_hatch_round = False\n",
        "            forth_hatch_round = True\n",
        "            continue\n",
        "        elif color_idx == NUM_COLORS-1 and forth_hatch_round:\n",
        "            hatch = \"+\"\n",
        "            color_idx = 0\n",
        "            continue\n",
        "        else:\n",
        "            color_idx += 1\n",
        "\n",
        "# Add number of total virus counts to plot and generate xtick labels\n",
        "ax.set_xlim(0, 1.08)\n",
        "sample_type_label = []\n",
        "x_height  = 1.005\n",
        "for i, sample in enumerate(samples):\n",
        "    total_count = \"{:,}\".format(vir_fractions_df[vir_fractions_df.index == sample][\"sample_total\"].values[0].astype(int))  \n",
        "    ax.annotate(total_count, (x_height, i), fontsize=fontsize-1, ha=\"left\", va=\"center\")\n",
        "    \n",
        "    sample_type_label.append(sample)\n",
        "\n",
        "# ax.set_yscale(\"log\")\n",
        "ax.set_xlabel(\"Fraction of total virus reads\", fontsize=fontsize)\n",
        "ax.set_ylabel(\"Animal\", fontsize=fontsize)\n",
        "\n",
        "# leg = ax.legend(fontsize=fontsize, loc='center left', bbox_to_anchor=(1, 0.5))\n",
        "leg = ax.legend(fontsize=fontsize-3, loc='upper center', bbox_to_anchor=(0.5, -0.05), ncol=7)\n",
        "for i, leg_txt in enumerate(leg.get_texts()):\n",
        "    leg_txt.set_color(\"white\")\n",
        "leg.get_frame().set_alpha(0.2)\n",
        "\n",
        "ax.set_yticks(x, sample_type_label, rotation=0, ha=\"right\")\n",
        "        \n",
        "ax.tick_params(axis=\"both\", labelsize=fontsize)\n",
        "ax.set_title(f\"Fraction of reads per virus species\", fontsize=fontsize+2)\n",
        "    \n",
        "ax.margins(y=0.01) \n",
        "\n",
        "fig.show()"
      ],
      "metadata": {
        "id": "eogXoitaPNVI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Create [Krona](https://github.com/marbl/Krona/wiki) plot \n",
        "This plot will be saved as html and can be opened in a browser."
      ],
      "metadata": {
        "id": "gQyDHROJPVxS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install Krona from source\n",
        "!git clone https://github.com/marbl/Krona.git && cd Krona && ./install.pl"
      ],
      "metadata": {
        "id": "_RoF_yapQUWu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "master = pd.DataFrame()\n",
        "for idx, timepoint in enumerate(adata.obs[\"dpi_clean\"].unique()):\n",
        "    for animal_id in adata.obs[adata.obs[\"dpi_clean\"] == timepoint][\"donor_animal\"].unique():\n",
        "        adata_temp = adata[(adata.obs[\"dpi_clean\"] == timepoint) & (adata.obs[\"donor_animal\"] == animal_id), :]\n",
        "\n",
        "        # Add total number of counts (across all cells) for each virus ID to phylogeny data temp\n",
        "        virus_ids = adata_temp.var.index.values\n",
        "        total_counts = nd(adata_temp.X.sum(axis=0))\n",
        "        total_count_dict = {virus_ids[i]: total_counts[i] for i in range(len(virus_ids))}\n",
        "\n",
        "        phylogeny_data_temp = phylogeny_data.copy()\n",
        "        phylogeny_data_temp['total_count'] = phylogeny_data_temp['Label'].map(total_count_dict)\n",
        "\n",
        "        # Remove all viruses that were only detected in less than num_cells_threshold cells\n",
        "        phylogeny_data_temp = phylogeny_data_temp[phylogeny_data_temp[\"total_count\"] >= num_cells_threshold]\n",
        "\n",
        "        # Remove non-relevent columns and change order of columns  \n",
        "        phylogeny_data_temp = phylogeny_data_temp[[\"total_count\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\", \"Label\"]]\n",
        "\n",
        "        # Replace dots with NaN\n",
        "        phylogeny_data_temp = phylogeny_data_temp.replace(\".\", np.nan)\n",
        "\n",
        "        # Add column with timepoint\n",
        "        phylogeny_data_temp[\"timepoint\"] = timepoint\n",
        "\n",
        "        # Add column with animal id\n",
        "        phylogeny_data_temp[\"animal_id\"] = animal_id\n",
        "\n",
        "        # Append to master dataframe\n",
        "        if idx == 0:\n",
        "            master = phylogeny_data_temp.copy()\n",
        "        else:\n",
        "            master = master.append(phylogeny_data_temp)\n",
        "\n",
        "# Save counts + taxnomomies data to txt file\n",
        "master.to_csv(f'krona.txt', sep ='\\t', header=None, index=False)\n",
        "\n",
        "# Generate Krona plot\n",
        "krona_out = \"figures/krona.html\"\n",
        "!$ktImportText krona.txt -o $krona_out -n \"Viral RNAs\""
      ],
      "metadata": {
        "id": "cNfWG2_yPTkO"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}